Implemented Near Cache configuration to store its keys by-reference

Donnerbart opened the pull request:
* added `serializeKeys` to `NearCacheConfig`, `NearCacheConfigReadOnly`,
  `XmlConfigBuilder`, `XmlClientConfigBuilder` and XSD files
* added `isSerializeKeys()` to `NearCache`
* added `setSerializeKeys()` to `NearCacheTestUtils.createNearCacheConfig()`
* added on-demand key de-serialization to `RepairingHandler` and
  `NearCacheTestUtils`

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/8424/

Donnerbart commented:
run-lab-run

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/8427/

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/8429/

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/8464/

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/8712/

Donnerbart commented:
run-lab-run

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/8740/

Donnerbart commented:
run-lab-run

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/8896/

Donnerbart commented:
run-lab-run

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

Donnerbart commented on hazelcast-spring/src/main/resources/hazelcast-spring-3.9.xsd:
```diff
@@ -2527,7 +2527,8 @@
         <xs:attribute name="in-memory-format" type="in-memory-format" use="optional" default="BINARY"/>
-        <xs:attribute name="invalidate-on-change" use="optional" type="xs:boolean" default="true"/>
+        <xs:attribute name="serialize-keys" use="optional" type="parameterized-boolean" default="true"/>
+        <xs:attribute name="invalidate-on-change" use="optional" type="parameterized-boolean" default="true"/>
```
This PR also fixes the type of `invalidate-on-change` to `parameterized-boolean`.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/8956/

Donnerbart commented:
run-lab-run

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/8986/

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.

Donnerbart commented:
run-lab-run

devOpsHazelcast commented:
Test PASSed.

vbekiaris commented on hazelcast/src/main/java/com/hazelcast/config/NearCacheConfig.java:
```diff
@@ -92,6 +92,7 @@
 
     private String name = "default";
     private InMemoryFormat inMemoryFormat = DEFAULT_MEMORY_FORMAT;
+    private boolean serializeKeys = true;
```
Is this the correct default value? Or is it updated in another PR?

Donnerbart replied:
I'm not sure if all code parts will work until the whole serialization implementation is not in place, so for the first step is to keep the old behavior (serialize keys). Then the data structure implementations should follow. After that the `StaleReadDetector` fixes. In the end we can switch the default to `false` and all tests etc. should still work :)

devOpsHazelcast commented:
Test PASSed.

Donnerbart commented:
I just added the basis to deal with `serializeKeys` and `NATIVE` memory format. Since the HD Near Cache cannot store keys by-reference, I simply override the value in the configuration. I've added documentation to the XSD and `NearCacheConfig` JavaDoc. Once we support this, we simply can remove that override.

The `XmlConfigBuilder` will also log a warning if the option was explicitly set.

Donnerbart commented on hazelcast/src/main/java/com/hazelcast/config/NearCacheConfig.java:
```diff
@@ -261,9 +263,35 @@ public NearCacheConfig setInMemoryFormat(String inMemoryFormat) {
+     * or in-memory-format is {@code NATIVE}
+     */
+    public boolean isSerializeKeys() {
+        return serializeKeys && inMemoryFormat != InMemoryFormat.NATIVE;
```
Here is the override to disable the feature for `NATIVE` in-memory-format.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/9209/

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/9210/

devOpsHazelcast commented:
Test PASSed.

devOpsHazelcast commented:
Test PASSed.
